src/manifolds/Segre.jl

```diff
@@ -0,0 +1,473 @@
+        x_1 \cos\mathopen{\Big(} \frac{g \lVert u_1 \rVert_{x_1}}{m} \mathclose{\Big)} + \frac{u_1}{\lVert u_1 \rVert_{x_1}} \sin\mathopen{\Big(} \frac{g \lVert u_1 \rVert_{x_1}}{m} \mathclose{\Big)},\\
+        \dots,\\
+        x_d \cos\mathopen{\Big(} \frac{g \lVert u_d \rVert_{x_d}}{m} \mathclose{\Big)} + \frac{u_d}{\lVert u_d \rVert_{x_d}} \sin\mathopen{\Big(} \frac{g \lVert u_d \rVert_{x_d}}{m} \mathclose{\Big)}
+    \right),
+````
+where
+````math
+    g = \tan^{-1}\mathopen{\Big(} t \frac{\sqrt{P^2 + 1}}{\lambda} + P \mathclose{\Big)} - \tan^{-1}(P),\\
+    m = \sqrt{\lVert u_1 \rVert_{x_1}^2 + \dots + \lVert u_d \rVert_{x_d}^2},\\
+    P = \frac{\nu}{\lambda m},\\
+    t = \lVert v \rVert_{p}.
+````
+If ``m = 0`` and ``\nu t < \lambda``, then ``\operatorname{exp}_p(v) = p + v``.
+
+For a proof, see proposition 3.1 in [JacobssonSwijsenVandervekenVannieuwenhoven:2024](@cite).
+"""
+exp(M::Segre{ℝ,V}, p, v) where {V}
+
+function exp!(M::Segre{ℝ,V}, q, p, v) where {V}
+    m_ = sqrt(
+        sum([
+            norm(Sphere(n - 1), x, xdot)^2 for (n, x, xdot) in zip(V, p[2:end], v[2:end])
+        ]),
+    )
+
+    q[1][1] = sqrt((p[1][1] + v[1][1])^2 + (p[1][1] * m_)^2)
+
+    f = pi / 2 - atan((p[1][1] + v[1][1]) / (p[1][1] * m_))
+    if m_ == 0
+        for (x, y) in zip(p[2:end], q[2:end])
+            y .= x
+        end
+    else
+        for (n, x, y, xdot) in zip(V, p[2:end], q[2:end], v[2:end])
+            a = norm(Sphere(n - 1), x, xdot)
+            y .= x * cos(a * f / m_) .+ xdot * (f / m_) * sinc(a * f / (m_ * pi))
+        end
+    end
+
+    return 0
+end
+
+@doc raw"""
+    function log(M::Segre{ℝ, V}, p, q)
+
+Logarithmic map on the Segre manifold.
+
+Let ``p \doteq (\lambda, x_1, \dots, x_d)``, ``q \doteq (\mu, y_1, \dots, y_d) \in \mathcal{S}``.
+Assume ``p`` and ``q`` are connected by a geodesic.
+Let
+````math
+    m = \sqrt{\sphericalangle(x_1, y_1)^2 + \dots + \sphericalangle(x_d, y_d)^2}
+````
+and assume ``(\mu, y_1, \dots, y_d)`` is the representation of ``q`` that minimizes ``m``. Then
+````math
+    \operatorname{log}_p(q) =
+    c \left(
+        \frac{\lambda m \mathopen{\Big(} \operatorname{cos}(m) - \frac{\lambda}{\mu} \mathclose{\Big)}}{\operatorname{sin}(m)},
+        \frac{\sphericalangle(x_1, y_1) (y_1 - \langle x_1, y_1 \rangle x_1)}{\sin(\sphericalangle(x_1, y_1))},
+        \dots,
+        \frac{\sphericalangle(x_d, y_d) (y_d - \langle x_d, y_d \rangle x_d)}{\sin(\sphericalangle(x_d, y_d))}
+    \right),
+````
+where ``c`` is determined by ``\lVert \operatorname{log}_p(q) \rVert_{p} = \operatorname{dist}(p, q)``.
+
+For a proof, see theorem 4.4 in [JacobssonSwijsenVandervekenVannieuwenhoven:2024](@cite).
+"""
+function log(M::Segre{ℝ,V}, p, q) where {V}
+    q_ = closest_representation(M, p, q)
+    if connected_by_geodesic(M, p, q_)
+        v = zeros.(size.(p)) # Initialize
+        log!(M, v, p, q_)
+        return v
+    else
+        return Nothing
```

Comment:
This seems a bit random and not very type stable.
Also please implement.`log!` (the same way you nicely implemented `exp!`) instead.

We could fill the result with `NaN` if it does not exist or so instead?

Reply:
Hi, I noticed now that even if `log!` fills `v` with nans, it still gets interpreted as floats in `log`. So it outputs for example something like `[[4.0e-323], [6.9057512179161e-310, 6.90574352987124e-310], [6.90574432576356e-310, 6.90570212344215e-310], [-0.0245041404219386, 0.002821512911371685], [-0.01594663570864798, -0.01842389578578475], [-0.06178003383876796, -0.0713773692384812]]`. So I don't know what the best way to deal with this is. Maybe to put the check in `log` instead?

Reply:
Well, `NaN`s are floats. So of course that is the case. That is intentional – the `Nothing` this comment is based on would be type-unstable. And I would not overwrite `log`

What would you wish for to return? 

As an example – `log`/`log!` on the sphere does not to any checks whether from `M, p, q` the `q` is on the sphere (unit norm). If it is not, the result is simply not defined and you might get out _anything_. The same for opposing points.
I would do the same here. If it is not well-defined, just don't care what you return.
Actually also on the sphere, your new function `connected_by_geodesic` might also be helpful – maybe with a keyword argument to ask whether the shortest geodesic is unique (because only then then log would be well-defined for sure)

Reply:
Ah, I didn't realize nans are floats as well. Then it must be something else then... I replaced the `return Nothing` above with `return v = NaN * p`, and I was confused by this behaviour:
```
julia> M = Segre(2, 2, 2, 2, 2);

julia> p = [[1.0], [1.0, 0.0], [1.0, 0.0], [1.0, 0.0], [1.0, 0.0], [1.0, 0.0]];

julia> q = [[1.0], [0.0, 0.1], [0.0, 1.0], [0.0, 1.0], [0.0, 1.0], [0.0, 1.0]];

julia> v = deepcopy(p);

julia> log!(M, v, p, q)
6-element Vector{Vector{Float64}}:
 [NaN]
 [NaN, NaN]
 [NaN, NaN]
 [NaN, NaN]
 [NaN, NaN]
 [NaN, NaN]

julia> log(M, p, q)
6-element Vector{Vector{Float64}}:
 [5.0e-324]
 [6.365987374e-314, 1.0e-323]
 [6.93009044731883e-310, 6.93009779249477e-310]
 [6.9300904323103e-310, 6.9300947100619e-310]
 [6.93008946239225e-310, 6.9300977939335e-310]
 [6.93009783360024e-310, 6.9300978225541e-310]

julia> 
```
In my experience, the e-310's usually indicate that I tried to interpret something that isnt a float as a float, which is why I assumed that's what had happened. But if you don't usually do this check on the input, maybe I'll just remove it :).

Reply:
Maybe two points

1. is it really necessary to fill it with nans? Does the other code (if the geodesic exists) error if called with wrong input? Or does it just fill the result with nonsense? If the second is the case, we can also leave it at that nonsense (instead of nans)
2. The current form allocates quite a bit, I think `v .= NaN .* p` should be better, maybe there is even a better `fill!.(v, NaN)` possible (both ideas, not tested).

the `e-310` just mean that you allocated memory for `v` but never filled it with any values also not with `NaN`s. The thing I like less here, is that `log` and `log!` return different results, that should never be the case.

Reply:
yea just returning nonsense is easier, let's do that.

I have no idea why `log` and `log!` were returning different results, it's possible it was because `log!` was not also returning `v`, which I fixed now, but I don't know